## Типовий план EDA (Exploratory Data Analysis / розвідувального аналізу даних) 👇

# 1. Розуміння даних

Перегляд перших/останніх рядків (head, tail)

Розмір датасету (shape)

Типи змінних (info)

Опис колонок (бізнес-сенс)

# 2. Очищення даних

Видалення дублікатів

Обробка пропусків (NaN)

Фільтрація некоректних значень (нульові/від’ємні ціни, кількості)

Перевірка аномалій

# 3. Описова статистика

Середнє, медіана, мін/макс

Розподіли числових змінних

Частоти категоріальних змінних

# 4. Аналіз окремих змінних (Univariate)

Гістограми

Boxplot (викиди)

Частотні таблиці

# 5. Аналіз взаємозв’язків (Bivariate / Multivariate)

Кореляції

Scatter plots

Залежність цільової змінної від факторів

Групування (groupby)

# 6. Виявлення патернів

Сезонність / тренди (якщо є дата)

Поведінка клієнтів / товарів

Найприбутковіші категорії

# 7. Перевірка гіпотез

Порівняння груп

Логічні припущення з даних

Підготовка до статистичних тестів / ML

# 8. Висновки

Ключові інсайти

Проблеми в даних

Що робити далі (ML, сегментація, прогноз)

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime as dt
import seaborn as sns

In [2]:
# Приклад
df = pd.read_excel('Online Retail.xlsx') # !! треба змінити на свій шлях !!

In [3]:
df.shape

(541909, 8)

In [4]:
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [5]:
df.head(2)

InvoiceNo StockCode                         Description  Quantity  \
0    536365    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                 WHITE METAL LANTERN         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [6]:
# EDA
df.info()
# бачимо типи даних та кількість не-null значень у кожному стовпці
# Виявляємо, що деякі стовпці мають пропущені значення (Descriptiom CustomerID), 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [8]:
df.isnull().sum()
# бачимо кількість пропущених значень у кожному стовпці

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [7]:
df.describe()
# отримуємо основні статистичні показники для числових стовпців
# бачимо, що мінімальні значення Quantity та UnitPrice від'ємні 

Quantity                    InvoiceDate      UnitPrice  \
count  541909.000000                         541909  541909.000000   
mean        9.552250  2011-07-04 13:34:57.156386048       4.611114   
min    -80995.000000            2010-12-01 08:26:00  -11062.060000   
25%         1.000000            2011-03-28 11:34:00       1.250000   
50%         3.000000            2011-07-19 17:17:00       2.080000   
75%        10.000000            2011-10-19 11:27:00       4.130000   
max     80995.000000            2011-12-09 12:50:00   38970.000000   
std       218.081158                            NaN      96.759853   

          CustomerID  
count  406829.000000  
mean    15287.690570  
min     12346.000000  
25%     13953.000000  
50%     15152.000000  
75%     16791.000000  
max     18287.000000  
std      1713.600303

In [ ]:
df[df['Quantity'] < 0].shape
# бачимо кількість рядків з від'ємними значеннями Quantity

(10624, 8)

In [ ]:
df[df['UnitPrice'] < 0].shape
# бачимо кількість рядків з від'ємними значеннями UnitPrice

(2, 8)

In [ ]:
df['CustomerID'].nunique()
# унікальні клієнти

4372

In [ ]:
df['Country'].unique()
# унікальні країни
# Бачимо, що є країна "Unspecified", можливо, варто розглянути, як з нею працювати далі

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [9]:
# виявлення дублікатів
df.duplicated().sum()

np.int64(5268)

# ВИСНОВКИ

(541909, 8) розмірність датесету, 8-мь колонок

5 268 дублікатів

стовпці Descriptiom, CustomerID мають пропущені значення 

значення 10 624 Quantity та 2  UnitPrice від'ємні

необхідно видалити дублікати, та здійснити фільтровку данних,  залишаючи лише ті рядки, де Quantity та UnitPrice більші за 0, а CustomerID не є порожнім (null)


In [10]:
# Data cleaning
df.drop_duplicates(inplace=True)
# Видаляє дублікати рядків у DataFrame.
# inplace=True означає, що зміни застосовуються безпосередньо до df, без створення нової копії

df = df[(df['Quantity'] > 0) & (df['UnitPrice'] > 0) & (df['CustomerID'].notnull())]
# Фільтрує дані, залишаючи лише ті рядки, де Quantity та UnitPrice більші за 0, а CustomerID не є порожнім (null).

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 392692 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    392692 non-null  object        
 1   StockCode    392692 non-null  object        
 2   Description  392692 non-null  object        
 3   Quantity     392692 non-null  int64         
 4   InvoiceDate  392692 non-null  datetime64[ns]
 5   UnitPrice    392692 non-null  float64       
 6   CustomerID   392692 non-null  float64       
 7   Country      392692 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 27.0+ MB


In [13]:
df.duplicated().sum()

np.int64(0)